errors/errors.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package errors
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+const (
+	ErrorCodesFeatureMetadataKey = "error_codes_feature"
+	Owner                        = "components-contrib"
+	Domain                       = "dapr.io"
+	unknown                      = "UNKNOWN_REASON"
+)
+
+const (
+	// gRPC to HTTP Mapping: 500 Internal Server Error
+	unknownHTTPCode = http.StatusInternalServerError
+)
+
+var UnknownErrorReason = WithErrorReason(unknown, unknownHTTPCode, codes.Unknown)
+
+type ResourceInfo struct {
+	Type string
+	Name string
+}
+
+// call this function to apply option
+type Option func(*Error)
+
+type Error struct {
```

Comment:
Please add comments explaining what these types are. As I understand it is a structured error type which helps to send the correct status codes and context over gRPC and HTTP, but that's not clear unless you read the whole file.